src/wallet/wallet.cpp

```diff
@@ -79,6 +79,15 @@ std::shared_ptr<CWallet> GetWallet(const std::string& name)
     return nullptr;
 }
 
+// Custom deleter for shared_ptr<CWallet>.
+static void ReleaseWallet(CWallet* wallet)
+{
+    LogPrintf("Releasing wallet %s\n", wallet->GetName());
+    wallet->BlockUntilSyncedToCurrentChain();
```

Comment:
What's the reason for this call? It seems very fragile because BlockUntilSyncedToCurrentChain may try to schedule a callback, but wallets currently get deleted after the scheduler is deleted:

https://github.com/bitcoin/bitcoin/blob/baf3a3ab0c63b512b37d9f753768c1f020369088/src/init.cpp#L297-L299

I was seeing this cause a crash due to a subtle change in behavior in one of my prs, which took me a long time to debug.

Reply:
The initial idea was to avoid dangling pointers to the wallet in the validation queue. I agree that once in the deleter no more pointers to the wallet should exist. Do you suggest removing this call? Which PR are you referring?

Reply:
> The initial idea was to avoid dangling pointers to the wallet in the validation queue.

@promag I think it would be better to just call UnregisterValidationInterface in this case. That looks like it would remove any wallet references from the notification code without needing to block wallet deletion on the scheduler thread or run additional notification handling code.

>  I agree that once in the deleter no more pointers to the wallet should exist. Do you suggest removing this call? Which PR are you referring?

Yesterday the problem I saw was in #10973. It slightly changed the way `m_last_block_processed` got set during rescans leading to a new `SyncWithValidationInterfaceQueue()` call in `SyncWithValidationInterfaceQueue` that doesn't happen in master. The `SyncWithValidationInterfaceQueue` call would fail with an assert error due to signal scheduler being deleted before the wallet in the init code shown above. 

Today this is causing another problem in #11625. In this case, there seems to be a race condition there where the wallet sometimes gets deleted in a qt event thread after `UnloadBlockIndex` gets called in another thread.